modules/standard/Set.chpl

```diff
@@ -541,9 +541,12 @@ module Set {
     :return: A new set containing the union between `a` and `b`.
     :rtype: `set(?t, ?)`
   */
-  proc |(const ref a: set(?t, ?), const ref b: set(t, ?)): set(t) {
+  proc |(const ref a: set(?t, ?), const ref b: set(t, ?)) {
     var result: set(t, (a.parSafe || b.parSafe));
 
+    // TODO: Split-init causes weird errors.
```

Comment:
Let's say I remove the `result;` from `proc |`. I see output like this:

```
$CHPL_HOME/modules/standard/Set.chpl:544: In function '|':
$CHPL_HOME/modules/standard/Set.chpl:547: error: const actual is passed to a 'ref' formal of init=()
  setCompositionParSafe.chpl:57: called as |(a: set(int(64),false), b: set(int(64),false)) from function 'doTest'
  setCompositionParSafe.chpl:68: called as doTest(param mode = LockMode.neither, op: string) from function 'testSetCompositionOps'
  setCompositionParSafe.chpl:77: called as testSetCompositionOps(param mode = LockMode.neither)
$CHPL_HOME/modules/standard/Set.chpl:544: In function '|':
$CHPL_HOME/modules/standard/Set.chpl:547: error: const actual is passed to a 'ref' formal of init=()
  setCompositionParSafe.chpl:57: called as |(a: set(int(64),true), b: set(int(64),false)) from function 'doTest'
  setCompositionParSafe.chpl:68: called as doTest(param mode = LockMode.left, op: string) from function 'testSetCompositionOps'
  setCompositionParSafe.chpl:84: called as testSetCompositionOps(param mode = LockMode.left)$CHPL_HOME/modules/standard/Set.chpl:198: In method 'init=':
$CHPL_HOME/modules/standard/Set.chpl:200: error: Cannot replace an instantiated param field with another value
$CHPL_HOME/modules/standard/Set.chpl:200: note: field 'parSafe' has value 'true' but is set to 'false'
  $CHPL_HOME/modules/standard/Set.chpl:720: called as (set(int(64),true)).init=(other: set(int(64),false)) from function '^'
  setCompositionParSafe.chpl:61: called as ^(a: set(int(64),true), b: set(int(64),false)) from function 'doTest'
  setCompositionParSafe.chpl:68: called as doTest(param mode = LockMode.left, op: string) from function 'testSetCompositionOps'
  setCompositionParSafe.chpl:84: called as testSetCompositionOps(param mode = LockMode.left)
```

The first two errors don't make sense because `init=` has the type:

```chapel
proc init=(const ref other: set(?t, ?)) lifetime this < other
```

The last error I can probably fix with some fiddling? Not sure.

There might be a test case for `init=` with `set(?t, parSafe=true)` that I can start from, or I can wait to fiddle with these in another PR (which is why I added the future).

Reply:
The constness errors seem very strange (and probably related to compiler bugs and if not require more investigation) so fiddling with them in another PR is better.

The "Cannot replace an instantiated param field with another value" error is something you can fix in the `init=`. See for example https://github.com/chapel-lang/chapel/blob/fc3dbcc1d270e5d025671cda0529afbb9bf8f8ea/modules/internal/SharedObject.chpl#L283-L286

In your case it should probably be `if this.type.parSafe != ? then this.type.parSafe else other.parSafe` and similar for `eltType`.